File: ios/Sources/GearanCore/GearanSecureSession.swift
import CryptoKit
import Foundation
import Security

/// Ephemeral X25519 session. Wire public keys are RFC 7748 raw 32-byte values.
/// iOS supports only X25519 here; callers must not advertise optional P-256.
public final class GearanSecureSession {
    public enum Role { case initiator, responder }

    public let role: Role
    public private(set) var localPrivate: Curve25519.KeyAgreement.PrivateKey?
    public private(set) var peerPublicRaw: Data?
    public private(set) var nonceLocal: Data
    public private(set) var noncePeer: Data?
    public private(set) var sessionKey: SymmetricKey?
    public private(set) var sasCode: String?
    private var transcript = Data()

    public init(role: Role = .initiator) {
        self.role = role
        nonceLocal = Self.randomNonce()
    }

    public func beginHandshake() {
        localPrivate = Curve25519.KeyAgreement.PrivateKey()
        nonceLocal = Self.randomNonce()
        peerPublicRaw = nil
        noncePeer = nil
        transcript = Data()
        sessionKey = nil
        sasCode = nil
    }

    public var localPublicRaw: Data? { localPrivate?.publicKey.rawRepresentation }

    public func recordTranscript(_ parts: Data...) { parts.forEach { transcript.append($0) } }
    public func transcriptData() -> Data { transcript }

    public func setPeer(publicRaw: Data, nonce: Data) throws {
        guard publicRaw.count == 32 else {
            throw PairingError(code: "HANDSHAKE_KEY", message: "X25519 public key must be exactly 32 raw bytes")
        }
        guard nonce.count == 16 else {
            throw PairingError(code: "HANDSHAKE_NONCE", message: "pairing nonce must be exactly 16 bytes")
        }
        _ = try Curve25519.KeyAgreement.PublicKey(rawRepresentation: publicRaw)
        peerPublicRaw = publicRaw
        noncePeer = nonce
    }

    /// Derives HKDF(shared || nonceI || nonceW, salt, "session") and SAS.
    public func deriveSession() throws -> String {
        guard let privateKey = localPrivate, let peerRaw = peerPublicRaw, let peerNonce = noncePeer else {
            throw PairingError(code: "HANDSHAKE", message: "missing handshake material")
        }
        let peerKey: Curve25519.KeyAgreement.PublicKey
        do {
            peerKey = try Curve25519.KeyAgreement.PublicKey(rawRepresentation: peerRaw)
        } catch {
            throw PairingError(code: "HANDSHAKE_KEY", message: "invalid X25519 public key")
        }
        let shared = try privateKey.sharedSecretFromKeyAgreement(with: peerKey)
        let sharedBytes = shared.withUnsafeBytes { Data($0) }
        let (nonceI, nonceW) = Self.orderedNonces(role: role, local: nonceLocal, peer: peerNonce)
        let derived = GearanLink.hkdf(
            ikm: sharedBytes + nonceI + nonceW,
            salt: Data("Gearan-Link-0.1".utf8),
            info: Data("session".utf8),
            length: 32
        )
        sessionKey = SymmetricKey(data: derived)
        let sas = GearanLink.sas(transcript: Data(SHA256.hash(data: transcript)))
        sasCode = sas
        return sas
    }

    public func commitWord(role: String) throws -> Data {
        guard let key = sessionKey else {
            throw PairingError(code: "HANDSHAKE", message: "session key not derived")
        }
        var authenticator = HMAC<SHA256>(key: key)
        authenticator.update(data: Data(("commit" + role).utf8))
        authenticator.update(data: Data(SHA256.hash(data: transcript)))
        return Data(authenticator.finalize())
    }

    public func reset() {
        localPrivate = nil
        peerPublicRaw = nil
        noncePeer = nil
        transcript = Data()
        sessionKey = nil
        sasCode = nil
    }

    static func orderedNonces(role: Role, local: Data, peer: Data) -> (Data, Data) {
        role == .initiator ? (local, peer) : (peer, local)
    }

    private static func randomNonce() -> Data {
        var bytes = [UInt8](repeating: 0, count: 16)
        precondition(SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes) == errSecSuccess)
        return Data(bytes)
    }
}

File: wear/app/src/main/java/com/gearan/watch/security/GearanSecureSession.kt
package com.gearan.watch.security

import com.gearan.watch.util.GearanLog
import java.security.PublicKey
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/** Watch-side handshake. Watch is protocol responder (`nonceW`). */
class GearanSecureSession {
    enum class Role { INITIATOR, RESPONDER }

    val role = Role.RESPONDER
    var ephemeral: CryptoUtils.Ephemeral? = null
        private set
    var peerEphemeralPub: PublicKey? = null
        private set
    var group = CryptoUtils.EcdhGroup.X25519
        private set
    var nonceLocal = CryptoUtils.randomBytes(16)
        private set
    var noncePeer: ByteArray? = null
        private set
    var sessionKey: ByteArray? = null
        private set
    var encKey: ByteArray? = null
        private set
    var macKey: ByteArray? = null
        private set
    var sasCode: String? = null
        private set
    private val transcriptSink = mutableListOf<ByteArray>()

    fun beginHandshake() {
        ephemeral = CryptoUtils.generateEphemeral()
        group = requireNotNull(ephemeral).group
        nonceLocal = CryptoUtils.randomBytes(16)
        peerEphemeralPub = null
        noncePeer = null
        sessionKey = null
        encKey = null
        macKey = null
        sasCode = null
        transcriptSink.clear()
        GearanLog.crypto("handshake begin (group=$group)")
    }

    /** Canonical 32-byte X25519 or 65-byte uncompressed X9.63 P-256. */
    fun localPublicBytes(): ByteArray {
        val local = requireNotNull(ephemeral) { "handshake not started" }
        return CryptoUtils.publicKeyToWire(local.keyPair.public, local.group)
    }

    fun recordTranscript(vararg parts: ByteArray) { transcriptSink.addAll(parts) }
    fun transcript(): ByteArray = transcriptSink.fold(ByteArray(0)) { all, part -> all + part }

    fun onPeerEphemeral(wirePublicKey: ByteArray, peerNonce: ByteArray, peerGroup: CryptoUtils.EcdhGroup) {
        check(peerGroup == group) { "ECDH group mismatch local=$group peer=$peerGroup" }
        require(peerNonce.size == 16) { "pairing nonce must be exactly 16 bytes" }
        peerEphemeralPub = CryptoUtils.publicKeyFromWire(wirePublicKey, peerGroup)
        noncePeer = peerNonce.copyOf()
    }

    fun deriveSession(): String {
        val local = requireNotNull(ephemeral) { "handshake not started" }
        val peer = requireNotNull(peerEphemeralPub) { "peer public key missing" }
        val peerNonce = requireNotNull(noncePeer) { "peer nonce missing" }
        val shared = CryptoUtils.ecdhShared(local.keyPair.private, peer, group)
        val (nonceI, nonceW) = orderedNonces(role, nonceLocal, peerNonce)
        val key = CryptoUtils.sessionKey(shared, nonceI, nonceW)
        sessionKey = key
        sasCode = CryptoUtils.sasFromTranscript(CryptoUtils.sha256(transcript()))
        CryptoUtils.appKeys(key).also { (enc, mac) -> encKey = enc; macKey = mac }
        GearanLog.crypto("session derived (group=$group), SAS ready for display")
        return requireNotNull(sasCode)
    }

    fun commitWord(roleLabel: String): ByteArray {
        val key = requireNotNull(sessionKey) { "session key not derived" }
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(key, "HmacSHA256"))
        return mac.doFinal(("commit" + roleLabel).toByteArray() + CryptoUtils.sha256(transcript()))
    }

    fun reset() {
        ephemeral = null
        peerEphemeralPub = null
        noncePeer = null
        sessionKey = null
        encKey = null
        macKey = null
        sasCode = null
        transcriptSink.clear()
    }

    internal fun orderedNonces(role: Role, local: ByteArray, peer: ByteArray) = when (role) {
        Role.INITIATOR -> local to peer
        Role.RESPONDER -> peer to local
    }
}
